crates/accelerate/src/inverse_cancellation.rs

```diff
@@ -0,0 +1,193 @@
+    self_inverse_gate_names: HashSet<String>,
+    self_inverse_gates: Vec<OperationFromPython>,
+) -> PyResult<()> {
+    if !self_inverse_gate_names
+        .iter()
+        .any(|name| op_counts.contains_key(name))
+    {
+        return Ok(());
+    }
+    for gate in self_inverse_gates {
+        let gate_count = op_counts.get(gate.operation.name()).unwrap_or(&0);
+        if *gate_count <= 1 {
+            continue;
+        }
+        let mut collect_set: HashSet<String> = HashSet::with_capacity(1);
+        collect_set.insert(gate.operation.name().to_string());
+        let gate_runs: Vec<Vec<NodeIndex>> = dag.collect_runs(collect_set).unwrap().collect();
+        for gate_cancel_run in gate_runs {
+            let mut partitions: Vec<Vec<NodeIndex>> = Vec::new();
+            let mut chunk: Vec<NodeIndex> = Vec::new();
+            let max_index = gate_cancel_run.len() - 1;
+            for (i, cancel_gate) in gate_cancel_run.iter().enumerate() {
+                let node = &dag.dag[*cancel_gate];
+                if let NodeType::Operation(inst) = node {
+                    if gate_eq(py, inst, &gate)? {
+                        chunk.push(*cancel_gate);
+                    } else {
+                        let is_empty: bool = chunk.is_empty();
+                        if !is_empty {
+                            partitions.push(std::mem::take(&mut chunk));
+                        }
+                        continue;
+                    }
+                    if i == max_index {
+                        partitions.push(std::mem::take(&mut chunk));
+                    } else {
+                        let next_qargs = if let NodeType::Operation(next_inst) =
+                            &dag.dag[gate_cancel_run[i + 1]]
+                        {
+                            next_inst.qubits
+                        } else {
+                            panic!("Not an op node")
+                        };
+                        if inst.qubits != next_qargs {
+                            partitions.push(std::mem::take(&mut chunk));
+                        }
+                    }
+                } else {
+                    panic!("Not an op node");
+                }
+            }
+            for chunk in partitions {
+                if chunk.len() % 2 == 0 {
+                    dag.remove_op_node(chunk[0]);
+                }
+                for node in &chunk[1..] {
+                    dag.remove_op_node(*node);
+                }
+            }
+        }
+    }
+    Ok(())
+}
+fn run_on_inverse_pairs(
+    py: Python,
+    dag: &mut DAGCircuit,
+    op_counts: &IndexMap<String, usize, RandomState>,
+    inverse_gate_names: HashSet<String>,
+    inverse_gates: Vec<[OperationFromPython; 2]>,
+) -> PyResult<()> {
+    if !inverse_gate_names
+        .iter()
+        .any(|name| op_counts.contains_key(name))
+    {
+        return Ok(());
+    }
+    for pair in inverse_gates {
```

Comment:
I suppose this is a faithful port first an foremost, but it might be nice to see the `inverse_gates` get unpacked to `gate_0` and `gate_1`, here.

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/13013/commits/00153dd8e0e9c1b6aeb3135d044c9d08d937bbc5